fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -1092,41 +1092,351 @@ inline bool tenantMapChanging(MutationRef const& mutation) {
 	return false;
 }
 
+// return the first tenantId whose idToPrefix(id) >= prefix[0..8] in lexicographic order. If no such id, return
+// INVALID_TENANT;
+inline int64_t lowerBoundTenantId(const StringRef& prefix, const std::map<int64_t, TenantName>& tenantMap) {
+	if (isSystemKey(prefix)) {
+		return TenantInfo::INVALID_TENANT;
+	}
+	Optional<int64_t> id;
+	if (prefix.size() >= TenantAPI::PREFIX_SIZE) {
+		// directly extract without copy
+		id = TenantAPI::extractTenantIdFromKeyRef(prefix);
+	} else {
+		// fall to generate a prefix
+		id = TenantIdCodec::lowerBound(prefix);
+	}
+	if (id.present() && id.get() >= 0) {
+		auto it = tenantMap.lower_bound(id.get());
+		if (it != tenantMap.end()) {
+			return it->first;
+		}
+	}
+	return TenantInfo::INVALID_TENANT;
```

Comment:
I think this could be written to rely more on `TupleIdCodec::lowerBound` as such:

```suggestion
	Optional<int64_t> id = TenantIdCodec::lowerBound(prefix.substr(0, std::min(prefix.size(), TenantAPI::PREFIX_SIZE)));
	if (id.present()) {
		auto it = tenantMap.lower_bound(id.get());
		if (it != tenantMap.end()) {
			return it->first;
		}
	}
	return TenantInfo::INVALID_TENANT;
```

Reply:
`TenantIdCodec::lowerBound(prefix.substr(0, std::min(prefix.size(), TenantAPI::PREFIX_SIZE)))` will return a negative number for keys like `"\x99\x01\x02\x03\x04\x05\x06\x07\x08"`. Then the lower_bound becomes 0, which get an ASSERTION failure in my unit test.

We still need the wrapper `lowerBoundTenantId` to handle some edge cases

Reply:
`TenantIdCodec::lowerBound` should return a negative number. There was a bug earlier where it did, but I believe Jon has fixed this now (if you don't have the fix, you may need to merge main)